lib/floe/workflow/runner/docker.rb

```diff
@@ -18,34 +18,95 @@ def run!(resource, env = {}, secrets = {})
-            secrets_file.write(secrets.to_json)
-            secrets_file.flush
+            secrets_file = create_secret(secrets)
+            env["_CREDENTIALS"] = "/run/secrets"
+          end
+
+          output = run_container(image, env, secrets_file)
+
+          [0, output]
+        ensure
+          secrets_file&.close!
+        end
+
+        def run_async!(resource, env = {}, secrets = {})
+          raise ArgumentError, "Invalid resource" unless resource&.start_with?("docker://")
 
-            params << [:e, "_CREDENTIALS=/run/secrets"]
-            params << [:v, "#{secrets_file.path}:/run/secrets:z"]
+          image = resource.sub("docker://", "")
+
+          secrets_file = nil
+          if secrets && !secrets.empty?
+            secrets_file = create_secret(secrets)
+            env["_CREDENTIALS"] = "/run/secrets"
           end
 
-          params << image
+          begin
+            container_id = run_container(image, env, secrets_file, :detached => true)
+          rescue
+            cleanup(container_id, secrets_file)
+            raise
+          end
 
-          logger.debug("Running docker: #{AwesomeSpawn.build_command_line("docker", params)}")
-          result = AwesomeSpawn.run!("docker", :params => params)
+          [container_id, secrets_file]
+        end
 
-          [result.exit_status, result.output]
-        ensure
+        def cleanup(container_id, secrets_file)
+          delete_container(container_id) if container_id
           secrets_file&.close!
         end
 
+        def running?(container_id)
+          inspect_container(container_id).first.dig("State", "Running")
+        end
+
+        def success?(container_id)
+          inspect_container(container_id).first.dig("State", "ExitCode") == 0
+        end
+
+        def output(container_id)
+          docker!("logs", container_id).output
+        end
+
         private
 
         attr_reader :network
+
+        def run_container(image, env, secrets_file, detached: false)
+          params  = ["run"]
+          params += detached ? [:detach] : [:rm]
+          params += [[:net, "host"]] if @network == "host"
+          params += env.map { |k, v| [:e, "#{k}=#{v}"] }
+          params << [:v, "#{secrets_file.path}:/run/secrets:z"] if secrets_file
+          params << image
+
+          logger.debug("Running docker: #{AwesomeSpawn.build_command_line("docker", params)}")
+
+          result = docker!(*params)
+          result.output
+        end
+
+        def inspect_container(container_id)
+          JSON.parse(docker!("inspect", container_id).output)
+        end
+
+        def delete_container(container_id)
+          docker!("rm", container_id)
+        rescue
+          nil
+        end
+
+        def create_secret(secrets)
+          secrets_file = Tempfile.new
+          secrets_file.write(secrets.to_json)
+          secrets_file.flush
+          secrets_file
+        end
+
+        def docker!(*params, **kwargs)
```

Comment:
Is `*params` correct?  Since we aren't splatting them below, and we are sending in a normal Array from the caller, I'm surprised this worked.

Reply:
Oh, no actually, we are only sending a normal Array in one case, but it's that case I'm surprised works.

Reply:
The other callers are doing e.g. `docker!("rm", container_id)` it is only `run_container` that builds up an array of params because that one is more dynamic.  I liked the way `docker!("inspect", container)` looked more than `docker!(["inspect", container_id])` for most of the cases.  And we're passing in `docker!(*params)` which is why it works

Reply:
Oh my mistake...I misread the run! method and didn't see it splatted the array as in came in.